library/bigip_asm_policy.py

```diff
@@ -0,0 +1,546 @@
+
+    def activate(self):
+        task = self.apply_policy_on_device()
+        if self.wait_for_task(task):
+            return True
+        else:
+            raise F5ModuleError('Apply policy task failed.')
+
+    def deactivate(self):
+        result = self.deactivate_policy_on_device()
+        if result:
+            return True
+        else:
+            raise F5ModuleError('Policy deactivation failed.')
+
+    def wait_for_task(self, task):
+        while True:
+            task.refresh()
+            if task.status in ['COMPLETED', 'FAILURE']:
+                break
+            time.sleep(1)
+        if task.status == 'FAILURE':
+            return False
+        if task.status == 'COMPLETED':
+            return True
+
+    def is_activated(self):
+        result = self.policy_active()
+        if self.client.check_mode:
+            return True
+        return result
+
+    def policy_active(self):
+        policy = self.return_policy()
+        if policy.active is True:
+            return True
+        else:
+            return False
+
+    def return_policy(self):
+        policies = self.policies_on_device()
+        for policy in policies:
+            if policy.name == self.want.name:
+                return policy
+
+    def policy_exists_on_device(self):
+        policy = self.return_policy()
+        if policy:
+            return True
+        else:
+            return False
+
+    def policies_on_device(self):
+        policies = self.client.api.tm.asm.policies_s.get_collection()
+        return policies
+
+    def upload_to_device(self):
+        self.client.api.tm.asm.file_transfer.uploads.upload_file(self.want.file)
+
+    def import_policy_to_device(self):
+        self.upload_to_device()
+        time.sleep(2)
+        name = os.path.split(self.want.file)[1]
+        result = self.client.api.tm.asm.tasks.import_policy_s.import_policy.create(name=self.want.name, filename=name)
+        return result
+
+    def apply_policy_on_device(self):
+        policy = self.return_policy()
+        link = {'link': policy.selflink}
+        result = self.client.api.tm.asm.tasks.apply_policy_s.apply_policy.create(policyReference=link)
+        return result
+
+    def deactivate_policy_on_device(self):
+        policy = self.return_policy()
+        policy.modify(active=False)
+        if policy.active is False:
+            return True
+        else:
+            return False
+
+    def create_policy_from_template_on_device(self):
+        result = self.client.api.tm.asm.tasks.import_policy_s.import_policy.create(
+            name=self.want.name, policyTemplateReference=self.want.template_link
+        )
+        return result
+
+    def create_policy_on_device(self):
+        result = self.client.api.tm.asm.policies_s.policy.create(name=self.want.name)
+        return result
+
+    def delete_policy_on_device(self):
+        policy = self.return_policy()
+        policy.delete()
+        if policy.exists():
```

Comment:
this existence should be checked in the `remove()` method.

Also, this module is misnamed. it should be `remove_from_device`